### PR TITLE
Improve console handling of sanitycheck and support regex matching

### PR DESCRIPTION
**Pull request by nashif**
This introduces the concept of harness into sanitycheck and add initial support
for the Console harness which can match output from tests and match it against
a defined regex in the testcase file.

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Comment by nashif** [comment]
recheck

File: scripts/sanitycheck [hunk 1]
```diff
@@ -2288,7 +2328,7 @@ def generate_coverage(outdir, ignores):
                 ["lcov", "--remove", coveragefile, i, "--output-file",
                  coveragefile],
                 stdout=coveragelog)
-        subprocess.call(["genhtml", "-output-directory",
+        subprocess.call(["genhtml", "--legend", "-output-directory",
```
Comment (andrewboie) [inline scripts/sanitycheck]: nit: this change could be in another patch, but I don't care that much.

File: scripts/sanitycheck [hunk 2]
```diff
@@ -339,14 +339,19 @@ class NativeHandler(Handler):
         self.returncode = 0
         self.set_state("running", {})
 
-    def output_reader(self, proc, harness):
+    def _output_reader(self, proc, harness):
```
Comment (andrewboie) [inline scripts/sanitycheck]: just curious why this method was renamed.

File: samples/application_development/external_lib/sample.yaml [hunk 3]
```diff
@@ -2,6 +2,10 @@ sample:
   name: External Library
 tests:
   test:
-    platform_whitelist: qemu_x86
```
Comment (andrewboie) [inline samples/application_development/external_lib/sample.yaml]: maybe a line in the commit message on why this whitelist was removed.

File: samples/philosophers/sample.yaml [hunk 4]
```diff
@@ -3,4 +3,13 @@ sample:
 tests:
   test:
     tags: samples
+    extra_args: "-DDEBUG_PRINTF=1"
     harness: console
+    harness_config:
+      type: multi_line
+      repeat: 10
+      regex:
+        - ".*STARVING.*"
```
Comment (andrewboie) [inline samples/philosophers/sample.yaml]: do the regexes need to be in this order or does it pass if it encounters any of these strings in indeterminate order?

File: samples/philosophers/sample.yaml [hunk 5]
```diff
@@ -1,15 +1,26 @@
 sample:
   name: Dining Philosophers
```
Comment (andrewboie) [inline samples/philosophers/sample.yaml]: We could extend this more (maybe in another PR). Philosophers through ifdefs lets you use several different kernel object types for synchronization, i think we currently only test with mutexes IIRC.

File: samples/mpu/mpu_stack_guard_test/sample.yaml [hunk 6]
```diff
@@ -4,6 +4,8 @@ tests:
   test:
     tags: apps
```
Comment (andrewboie) [inline samples/mpu/mpu_stack_guard_test/sample.yaml]: this particular sample really need to be converted into a proper test.
i think there's a bug open on it somewhere.
nothing to change in your patch here, just whining

File: samples/subsys/logging/kernel_event_logger/sample.yaml [hunk 7]
```diff
@@ -4,6 +4,10 @@ sample:
   name: Kernel event logger Sample
 tests:
   test:
-    platform_whitelist: qemu_x86
```
Comment (andrewboie) [inline samples/subsys/logging/kernel_event_logger/sample.yaml]: maybe a note in the commit message on why we can remove the whitelist

File: arch/posix/CMakeLists.txt [hunk 8]
```diff
@@ -1,6 +1,7 @@
 zephyr_cc_option_ifdef(CONFIG_LTO -flto)
 
 zephyr_compile_options(
+  -Wno-format-truncation
```
Comment (andrewboie) [inline arch/posix/CMakeLists.txt]: our policy for REVERTME patches is that we need to file a bug on it and put the bug ID in the commit message

File: samples/drivers/crypto/sample.yaml [hunk 9]
```diff
@@ -17,7 +17,8 @@ tests:
         - "\\[general\\] \\[INF\\] ccm_mode: CCM mode ENCRYPT - Match"
         - "\\[general\\] \\[INF\\] ccm_mode: CCM mode DECRYPT - Match"
   test-micro:
-    tags: micro
+    #FIXME fails in qemu
+    platform_exclude: qemu_nios2
```
Comment (andrewboie) [inline samples/drivers/crypto/sample.yaml]: this patch needs to be a REVERTME with a bug filed on it, this could be a symptom of something serious (or not)

File: samples/mpu/mpu_stack_guard_test/sample.yaml [hunk 10]
```diff
@@ -5,7 +5,8 @@ tests:
...
     harness: console
     # We do not handle segfaults here
-    platform_exclude: native_posix
+    arch_exclude: posix riscv32 nios2
+    platform_exclude: qemu_cortex_m3
```
Comment (andrewboie) [inline samples/mpu/mpu_stack_guard_test/sample.yaml]: this patch needs to be a REVERTME with a bug filed on it, this could be a symptom of something serious (or not)

File: samples/hello_world/sample.yaml [hunk 11]
```diff
@@ -13,6 +13,6 @@ common:
 tests:
   singlethread:
     extra_args: CONF_FILE=prj_single.conf
-    filter: not CONFIG_BT and not CONFIG_GPIO_SCH
+    filter: not CONFIG_BT and not CONFIG_GPIO_SCH and ASSERT == 0
```
Comment (andrewboie) [inline samples/hello_world/sample.yaml]: does this work? shouldn't this be "not CONFIG_ASSERT"?

Comment (andrewboie) [reply samples/hello_world/sample.yaml]: also please file a bug on this and link to the commit message of this patch

Comment (nashif) [reply samples/hello_world/sample.yaml]: it does not work as not CONFIG_ASSERT because ASSERT is being added by sanitycheck using -DCONFIG_ASSERT. I want to change this and make sanitycheck add kconfig overlays instead of defining this via cmake

File: samples/philosophers/sample.yaml (hunk 5 above)
Comment (nashif) [reply samples/philosophers/sample.yaml]: yes, good idea

File: arch/posix/CMakeLists.txt (hunk 8 above)
Comment (aescolar-ot) [reply arch/posix/CMakeLists.txt]: In which file/line is this warning appearing and with what gcc version?

File: samples/hello_world/sample.yaml (hunk 11 above)
Comment (aescolar-ot) [reply samples/hello_world/sample.yaml]: Regarding sanity check adding kconfig overlays: Would this be sufficient https://github.com/zephyrproject-rtos/zephyr/issues/5723?

File: scripts/sanity_chk/harness.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            self.regex = config.get('regex', [] )
+            self.repeat = config.get('repeat', 1)
+            self.ordered = config.get('ordered', True)
+
+class Console(Harness):
+
+    def handle(self, line):
+        if self.type == "one_line":
+            pattern = re.compile(self.regex[0])
+            if pattern.match(line):
```
Comment (aescolar-ot) [inline scripts/sanity_chk/harness.py]: As commented in IRC:
This will only match the regular expression if the line does not have any prefix or postfix, which for some targets may be the case (for example a tracing tool may add a time stamp to each line of a trace). It would be nice if there was a `.*` added at the beginning and end of the expression.
So the line above would become ```pattern = re.compile(".*" + self.regex[0] + ".*")``` and the for multi-line ```pattern = re.compile(".*" + r + ".*")```

Otherwise, it would be nice to have at some point a more generic Handler mechanism. So that we could easily define that a given board has a different handler. Then we could ensure that the stream that comes out of the handler into this harness would be already free of extra characters.

File: scripts/sanitycheck [hunk 13]
```diff
@@ -321,58 +327,72 @@ class NativeHandler(Handler):
...
+            self.returncode = proc.returncode
+            if proc.returncode != 0:
+                if self.returncode == 1:
                     out_state = "failed"
-                    break
+                else:
+                    out_state = "failed valgrind"
 
-        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-s", self.outdir], shell=True)
+        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-b", "-s", self.outdir], shell=True)
```
Comment (aescolar-ot) [inline scripts/sanitycheck]: Just a note: Most sample apps, are configured without COVERAGE and therefore will not account for this yet.
I guess when/if https://github.com/zephyrproject-rtos/zephyr/issues/5723 is implemented, we could use it to override their configuration. So that when run in sanitycheck, all tests and samples run in the POSIX arch would be compiled with coverage, instead of depending on the project configuration.

File: samples/mpu/mpu_stack_guard_test/sample.yaml (hunk 6 above)
Comment (nashif) [reply samples/mpu/mpu_stack_guard_test/sample.yaml]: agree, will address in another PR

File: samples/application_development/external_lib/sample.yaml (hunk 3 above)
Comment (nashif) [reply samples/application_development/external_lib/sample.yaml]: done

File: arch/posix/CMakeLists.txt (hunk 8 above)
Comment (nashif) [reply arch/posix/CMakeLists.txt]: done

File: samples/subsys/logging/kernel_event_logger/sample.yaml (hunk 7 above)
Comment (nashif) [reply samples/subsys/logging/kernel_event_logger/sample.yaml]: done

File: scripts/sanity_chk/harness.py (hunk 12 above)
Comment (andrewboie) [reply scripts/sanity_chk/harness.py]: would re.search() also accomplish this?

Comment (aescolar-ot) [reply scripts/sanity_chk/harness.py]: I think you are right

File: arch/posix/CMakeLists.txt (hunk 8 above)
Comment (andrewboie) [reply arch/posix/CMakeLists.txt]: It just says "Tracked in issue" without an ID

Comment (nashif) [reply arch/posix/CMakeLists.txt]: hmm, it was at the beginning of the line and was treated as comment, fixed now.

**Review by aescolar — APPROVED** [review]
Apart from this last commit message, this PR LGTM.
